### DISCUSSION TITLE
namemap alias handling causes providers to conflict with each other

### DISCUSSION THREAD

**Discussion by davidben** [openssl/openssl#23369] [General Discussion]
I was looking at how provider lookup works in 3.x, and noticed a weird behavior of the namemap machinery.

OpenSSL allows an algorithm to advertise multiple names. However, rather than implementing this directly, it https://github.com/openssl/openssl/pull/8967 implemented deep inside the namemap machinery. If provider1 implements an algorithm with names "A" and "B", both "A" and "B" will map to the same number in the namemap.

But the namemap is shared across providers! So if provider2 only names an algorithm "A", suddenly loading provider1 will make "B" also pick up provider2.A! If provider3 thinks "A" and "B" are distinct algorithms, it conflicts and you cannot load provider1 and provider3 at the same time. Whichever is loaded first wins and the other has an error. Moreover, this error is discovered late because of all the lazy-loading machinery. This effect persists even after providers are unloaded. If one loads a provider that links "A" and "B" together, "A" and "B" are forever linked in that context, even after the provider is unloaded.

This doesn't seem like the right behavior. Much more intuitive would be that `EVP_MD_fetch("A")` considers provider1.A, provider2.A, provider3.A, etc., without caring about the other names of those algorithms. That provider1.B happens to be the same object as provider1.A but provider3.B happens to be a distinct object should not matter.

This came about because I was curious what it'd take to remove all the mutable caches and switch providers to a simpler and more performant immutable-after-construction design. Having this kind of provider cross-talk makes efficient designs trickier.

**Comment by t8m** [comment]
This is not a bug. This was a conscious design decision. I'll let @levitte and @mattcaswell to explain why this design decision was made but I do not think this should be changed at this point.

**Comment by levitte** [comment]
The assumption we made, quite early on, was that all algorithm names are global and that aliases exist.

So re global names, you can't have RSA and RSA being two completely different algorithms, for example.  The implementations might differ a bit, but on a primitive level, they are assumed to be the same...

Furthermore, re aliases, this is partly due to this already being present in pre-3.0 OpenSSL (see things like `EVP_PKEY_base_id()`, but also partly to be able to support things like `EVP_PKEY_is_a(pk, "RSA")`, regardless of if the underlying implementation was fetched using "rsaEncryption", "RSA", or "1.2.840.113549.1.1.1".
It's true that we could have done this differently, by having the provider do aliases through implementation re-use, and give them the ability to provide a base type name.  The signature OSSL_ALGORITHM for RSA would then have looked like this:

``` C
    { "RSA", "provider=default", ossl_rsa_signature_functions },
    { "rsaEncryption", "provider=default", ossl_rsa_signature_functions },
    { "1.2.840.113549.1.1.1", "provider=default", ossl_rsa_signature_functions },
```

However, that doesn't answer the assumed global nature of algorithm names.  In the early discussions, we knew there would be exactly this sort of conflict.  What we figured was that then there is a conflict, and not just on an implementation level, and that (to take your example), the authors of provider1 and provider2 would have to have a conversation and resolve that conflict.

On a very human level, this is akin to two different inventors in the world inventing an algorithm BLARGH (my favorite fictituous algo name), completely different algorithms, and then both approaching NIST, or IETF, or whatever international standardization group.  I very much imagine that said group would tell those inventors pretty quickly that, "... er, there's a name conflict, we can't have that".

**Reply by levitte** [reply]
Does this make some sort of sense?

**Reply by davidben** [reply]
Yes, your system does inherently assume everyone broadly agrees what names mean, but wedging them together at the namemap level is not the right conclusion. That makes the system's behavior very unpredictable. For example, even without a conflict, an application may use nothing from a provider, but accidentally rely on that provider's presence because that's the only provider that has an alias.

The alternative signature OSSL_ALGORITHM for RSA seems also a false dichotomy. Your documentation already says that an OSSL_ALGORITHM has multiple names, and that providers are expected expected to declare the variations. There would be no need to specify each name separately, because you all already support the colon separation.
https://www.openssl.org/docs/man3.1/man3/OSSL_ALGORITHM.html#algorithm_names
https://www.openssl.org/docs/man3.1/man3/OSSL_ALGORITHM.html#On-the-subject-of-algorithm-names

Indeed providers have to provide all the names because they cannot rely on other providers (which may not be used in the application) to wedge the names together for them. But because namemap causes provider cross-talk, it's very easy to not notice this in testing. So providers may accidentally rely on this behavior, making their behavior unpredictable.

The unpredictability extends to OpenSSL, as it means subtle changes in how you resolve providers (especially given the current lazy-loading design) can break these accidental dependencies. Unpredictable behavior is harder to make backwards-compatible. A project that prioritizes backwards-compatibility should therefore place a premium on avoiding unpredictable behavior.

And, of course, these effects also exacerbate OpenSSL 3.x's performance regressions, by making it harder for you all to fix problematic implementation decisions. (I'd actually meant to prototype a faster version for you all over the holidays. When I noticed this behavior, I moved on to other projects.)

**Reply by levitte** [reply]
Oh!  Could it be that the aliasing in itself isn't the center of the issue, that it's more about *when* it happens?  'cause yeah, you're right, the names themselves are loaded whenever an algorithm is fetched, so the potential name conflicts are potentially discovered late.

Changing this is technically easy, libcrypto could simply query the OSSL_ALGORITHM arrays for each know operation when a provider is loaded, and add all the names from those.  It won't help the naming conflict one bit, but will at least make it apparent much more predictably.

**Reply by davidben** [reply]
That certainly doesn't help, but no. Predictability problems still occur because the author of provider A cannot assume provider B has been loaded. But if they only test with A and B together and forget to include some aliases, they won't realize that A does not work standalone.
